qualtran/testing.py

```diff
@@ -373,18 +390,91 @@ def check_bloq_example_decompose(bloq_ex: BloqExample) -> Tuple[BloqCheckResult,
     return BloqCheckResult.PASS, ''
 
 
-def assert_wire_symbols_match_expected(bloq: Bloq, expected_ws: List[str]):
-    """Assert a bloq's wire symbols match the expected ones.
+def assert_equivalent_bloq_example_counts(bloq_ex: BloqExample) -> None:
+    """Assert that the BloqExample has consistent bloq counts.
 
-    Args:
-        bloq: the bloq whose wire symbols we want to check.
-        expected_ws: A list of the expected wire symbols.
+    Bloq counts can be annotated directly via the `Bloq.build_call_graph` override.
+    They can be inferred from a bloq's decomposition. This function asserts that both
+    data sources are present and that they produce the same values.
+
+    If both sources are present, and they disagree, that results in a `FAIL`. If only one source
+    is present, an `UNVERIFIED` exception is raised. If neither are present, a `MISSING` result
+    is raised.
+
+    Returns:
+        None if the assertions are satisfied.
+
+    Raises:
+        BloqCheckException if any assertions are violated or unverifiable due to partial
+        or missing information.
     """
-    ws = []
-    regs = bloq.signature
-    for i, r in enumerate(regs):
-        # note this will only work if shape = ().
-        # See: https://github.com/quantumlib/Qualtran/issues/608
-        ws.append(bloq.wire_symbol(Soquet(i, r)).text)
+    bloq = bloq_ex.make()
+    generalizer = bloq_ex.generalizer
 
-    assert ws == expected_ws
+    has_manual_counts: bool
+    has_decomp_counts: bool
+    manual_counts = None
+    decomp_counts = None
+
+    if bloq.build_call_graph.__qualname__.startswith('Bloq.'):
+        has_manual_counts = False
+    else:
+        has_manual_counts = True
+        # TODO: don't let the author sneak in decomposition??
```

Comment:
What does this mean?

Reply:
I guess I could remove the TODO, but there's an inherent limitation in checking to see if a method is overridden to see if we have an independent data source. A bloq author could override the `build_call_graph` method but just use the default behavior of counting up bloqs in the decomposition. This cross-check would report success even though the two sources of data are the same.

Reply:
changed the todo to a block comment explaining the consideration.